plugins/techdocs/src/reader/components/useReaderState.ts

```diff
@@ -0,0 +1,335 @@
+    return 'CONTENT_STALE_READY';
+  }
+
+  // the build timed out, but the content is still stale
+  if (activeSyncState === 'BUILD_TIMED_OUT') {
+    return 'CONTENT_STALE_TIMEOUT';
+  }
+
+  // the build failed, but the content is still stale
+  if (activeSyncState === 'ERROR') {
+    return 'CONTENT_STALE_ERROR';
+  }
+
+  // seems like the content is up-to-date (or we don't know yet and the sync process is still evaluating in the background)
+  return 'CONTENT_FRESH';
+}
+
+/**
+ * The state of the synchronization task. It checks whether the docs are
+ * up-to-date. If they aren't, it triggers a build.
+ */
+type SyncStates =
+  /** Checking if it should be synced */
+  | 'CHECKING'
+
+  /** Building the documentation */
+  | 'BUILDING'
+
+  /** Finished building the documentation */
+  | 'BUILD_READY'
+
+  /** Building the documentation timed out */
+  | 'BUILD_TIMED_OUT'
+
+  /** No need for a sync. The content was already up-to-date. */
+  | 'UP_TO_DATE'
+
+  /** An error occurred */
+  | 'ERROR';
+
+type ReducerActions =
+  | {
+      type: 'sync';
+      state: SyncStates;
+      syncError?: Error;
+    }
+  | {
+      type: 'content';
+      content?: string;
+      contentLoading: boolean;
+      contentError?: Error;
+      contentReload: () => void;
+    }
+  | { type: 'navigate'; path: string };
+
+type ReducerState = {
+  /**
+   * The path of the current page
+   */
+  path: string;
+
+  /**
+   * The current sync state
+   */
+  activeSyncState: SyncStates;
+
+  /**
+   * If true, the content is downloading from the storage.
+   */
+  contentLoading: boolean;
+  /**
+   * The content that has been downloaded and should be displayed.
+   */
+  content?: string;
+  /**
+   * When called, the content is reloaded without refreshing the page.
+   */
+  contentReload?: () => void;
+  /**
+   * If true, the content is considered stale and should be refreshed by the user via a refresh or a navigation.
+   */
+  contentIsStale: boolean;
+
+  contentError?: Error;
+  syncError?: Error;
+};
+
+export function reducer(
+  oldState: ReducerState,
+  action: ReducerActions,
+): ReducerState {
+  const newState = { ...oldState };
+
+  switch (action.type) {
+    case 'sync':
+      newState.activeSyncState = action.state;
+      newState.syncError = action.syncError;
+
+      // whatever is stored as content, it can be considered as being stale
+      if (newState.activeSyncState === 'BUILD_READY') {
+        newState.contentIsStale = true;
+
+        // reload the content if this was the initial build OR the page was missing in the old version
+        if (!newState.content && newState.contentReload) {
+          newState.contentReload();
+
+          // eagerly mark the content to load to not get synchronization issues since
+          // the async hook behind contentReload() doesn't update the reducer instantly
+          // and might flash the "not found" page
+          newState.contentLoading = true;
+        }
```

Comment:
I guess this workaround is kind-of a flaw in my design. I decided to decouple the async actions of downloading the content and executing the sync because they are independent. The 1st one returns the `{content, loading, error}` state while the 2nd returns the `CHECKING->BUILDING->...` status. Both are used as independent inputs to a function that composes both states into one. This special "iff no content exists and the sync finished, then retry the content loading" case depends on both independent states and thus causes synchronization errors.

I solved it as you proposed with a new state `BUILD_READY_RELOAD` that handles this case. I added a kind-of-workaround with a `useRef()` to get access to `content` and `contentReload` without restarting the sync process whenever the content updates. It should only run once for now.

As a result of this change, it would be trivial to implement https://github.com/backstage/backstage/issues/4071 by just renaming `useAsync` of the sync task to `useAsyncRetry` and returning the retry function to the UI component.

Reply:
Awesome, thanks for elaborating!

While in theory it would still be possible to avoid the synchronization issues completely by lifting all state to the reducer, the updated approach works equally well. By having a clean implementation of the reducer now we enable further improvements of the <Reader> component as you mentioned 🚀